pty/ptytest/ptytest.go

```diff
@@ -7,116 +7,248 @@ import (
 	"io"
 	"os"
 	"os/exec"
-	"regexp"
 	"runtime"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 	"unicode/utf8"
 
 	"github.com/stretchr/testify/require"
+	"golang.org/x/xerrors"
 
 	"github.com/coder/coder/pty"
 )
 
-var (
-	// Used to ensure terminal output doesn't have anything crazy!
-	// See: https://stackoverflow.com/a/29497680
-	stripAnsi = regexp.MustCompile("[\u001B\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[a-zA-Z\\d]*)*)?\u0007)|(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PRZcf-ntqry=><~]))")
-)
-
 func New(t *testing.T) *PTY {
 	ptty, err := pty.New()
 	require.NoError(t, err)
 
-	return create(t, ptty)
+	return create(t, ptty, "cmd")
 }
 
 func Start(t *testing.T, cmd *exec.Cmd) (*PTY, *os.Process) {
 	ptty, ps, err := pty.Start(cmd)
 	require.NoError(t, err)
-	return create(t, ptty), ps
+	return create(t, ptty, cmd.Args[0]), ps
 }
 
-func create(t *testing.T, ptty pty.PTY) *PTY {
-	reader, writer := io.Pipe()
-	scanner := bufio.NewScanner(reader)
+func create(t *testing.T, ptty pty.PTY, name string) *PTY {
+	// Use pipe for logging.
+	logDone := make(chan struct{})
+	logr, logw := io.Pipe()
 	t.Cleanup(func() {
-		_ = reader.Close()
-		_ = writer.Close()
+		_ = logw.Close()
+		_ = logr.Close()
+		<-logDone // Guard against logging after test.
 	})
 	go func() {
-		for scanner.Scan() {
-			if scanner.Err() != nil {
-				return
-			}
-			t.Log(stripAnsi.ReplaceAllString(scanner.Text(), ""))
+		defer close(logDone)
+		s := bufio.NewScanner(logr)
+		for s.Scan() {
+			// Quote output to avoid terminal escape codes, e.g. bell.
+			t.Logf("%s: stdout: %q", name, s.Text())
 		}
 	}()
 
+	// Write to log and output buffer.
+	copyDone := make(chan struct{})
+	out := newStdbuf()
+	w := io.MultiWriter(logw, out)
+	go func() {
+		defer close(copyDone)
+		_, err := io.Copy(w, ptty.Output())
+		_ = out.closeErr(err)
+	}()
 	t.Cleanup(func() {
+		_ = out.Close
 		_ = ptty.Close()
+		<-copyDone
 	})
+
 	return &PTY{
 		t:   t,
 		PTY: ptty,
+		out: out,
 
-		outputWriter: writer,
-		runeReader:   bufio.NewReaderSize(ptty.Output(), utf8.UTFMax),
+		runeReader: bufio.NewReaderSize(out, utf8.UTFMax),
 	}
 }
 
 type PTY struct {
 	t *testing.T
 	pty.PTY
+	out *stdbuf
 
-	outputWriter io.Writer
-	runeReader   *bufio.Reader
+	runeReader *bufio.Reader
 }
 
 func (p *PTY) ExpectMatch(str string) string {
-	var buffer bytes.Buffer
-	multiWriter := io.MultiWriter(&buffer, p.outputWriter)
-	runeWriter := bufio.NewWriterSize(multiWriter, utf8.UTFMax)
+	p.t.Helper()
+
 	complete, cancelFunc := context.WithCancel(context.Background())
 	defer cancelFunc()
+
+	timeout := make(chan error, 1)
 	go func() {
+		defer close(timeout)
 		timer := time.NewTimer(10 * time.Second)
 		defer timer.Stop()
 		select {
 		case <-complete.Done():
 			return
 		case <-timer.C:
```

Comment:
Could we select this channel in the block below? Seems like we could reduce code!

Reply:
Doh, of course, that's such a good idea, why didn't I see it :joy:.